### ISSUE TITLE
use tree-sitter

### ISSUE THREAD

**Issue by tpapp** [JuliaEditorSupport/julia-emacs#174]
`tree-sitter` framework for incremental parsing of source code. The [Julia implementation](https://github.com/tree-sitter/tree-sitter-julia) is now, in my opinion, fairly mature.

I am asking for comments about replacing our ad-hoc regexp-based parsing mechanisms with it. Specifically, it would help with 

1. font locking, 
2. indentation,
3. code navigation,

resolving a host of issues. 

I am aware that it isn't perfect, as nothing is, but at least improvements would go to a repository that helps all Julia users, not just those who use Emacs.

**Comment by tpapp** [comment]
@FelipeLema: since you wrote an [indentation solution](https://codeberg.org/FelipeLema/tree-sitter-indent.el) using tree-sitter, I would appreciate your comments. I recall you mentioning it [a while ago](https://github.com/JuliaEditorSupport/julia-emacs/issues/6#issuecomment-696359544).

**Comment by FelipeLema** [comment]
I agree that currently maintaining the regex parser is taking a toll on code readability and maintanability, even considering people who are well versed in (Emacs) Lisp. I agree that tree-sitter is a tool that has many eyes on them ("it is production quality") and that end users have only good things to say about it.

However, I'd say it gets complicated for Julia. There's 2 big aspects to consider: how well the parser is maintained and how well it can be maintained (by the Julia community).

First, not only is the Julia tree-sitter parser not complete, but its _pace_ may not fit the requirements of the Julia community (incl tooling people & Julia devs). It didn't fit me particularly: I did write the tree-sitter indent, and when I used the Julia parser I found a bug. Then I reported it, debugged it, [proposed a fix](https://github.com/tree-sitter/tree-sitter-julia/pull/15) and I'm still waiting on the problem to be addressed. Can the tree-sitter parser be hijacked by the julia community so we can improve it at a quicker pace? I honestly don't know.

After that, I endend up writing a [Julia code formatter](https://codeberg.org/FelipeLema/julia-formatter.el) because I thought it would easier to do so rather than pushing for changes upstream in the Julia tree-sitter parser. The name of this tool I did is actually a misnomer because it parses the code in a Julia buffer and does operations on its AST _just as tree-sitter does_.

So this brings up the second concern about tree-sitter: it is written in another language (and right now we need to write fixes in JS for the Julia parser). Correct me if I'm wrong, but I believe that most of the Julia community understands the long-term problems of maintaining several languages for their workflows. I would personally rather avoid writing JS if possible (which is what I ended up doing in the paragraph above).

My recommendation is to use [CSTParser](https://github.com/julia-vscode/CSTParser.jl) (or even the parser that comes with Julia binaries) to parse the AST and to copy code from tree-sitter.el to handle the items mentioned at the [top entry of this discussion](https://github.com/JuliaEditorSupport/julia-emacs/issues/174#issue-1261576377). From my experience with the Julia code formatter I'd recommend using [DaemonMode](https://github.com/dmolina/DaemonMode.jl) for Emacs-Julia communication (to have little-to-no response delay) as using json-rpc, as LSP does, [may bring problems when used in Windows](https://codeberg.org/FelipeLema/julia-formatter.el/issues/1).

Using CSTParser may have a positive effect as it lowers the barriers for Julia _end users_ to participate in maintaining this package (kinda what the Racket community was betting on when they switched to ChezScheme).

All this being said, I want to note that I use tree-sitter in neovim on an everyday basis and that everything (except for the Julia tree-sitter parser) works A-OK.

**Comment by tpapp** [comment]
@FelipeLema: thanks for the detailed explanation (incidentally, would you consider giving the PR you mentioned a friendly ping? Perhaps it was just overlooked --- that happens).

Conceptually, I can think of the following components for the features we need:

1. a parsing framework, if applicable (eg tree-sitter, LSP)
2. Julia-specific part within that framework (eg tree-sitter-julia, LanguageServer.jl)
3. IDE-specific integration with the framework that calls this code, if applicable (emacs-tree-sitter, emacs-lsp)
4. a bit of mode-specific glue that calls this

Currently in julia-mode we pretty much do everything above with hacks using regexps.

CSTParser.jl does 1+2. We would have to maintain part 3 ourselves, most likely using a daemon-based approach you outline. The advantage is indeed doing a lot in Julia. The disadvantage is all the framework associated with maintaining a running instance of Julia --- doable but somewhat heavyweight.

Tree-sitter would allow us to combine effort for 1, 2, and 3 with other projects (editors other than Emacs, languages other than Julia). That said, I realize that a lot of layers can introduce problems, too.

Interestingly, the LSP spec includes semantic tokens since 3.16. I wonder if that's supported in practice for Julia with Emacs, @gdkrmr and @non-Jedi, it would be great if you could share your thoughts about this. If we could make that work, it would cover pretty much everything for us.

**Comment by dahtah** [comment]
Note that [julia-snail](https://github.com/gcv/julia-snail/) already includes an interface to CSTParser.jl. It still relies on julia-mode for syntax highlighting and formatting, though.

**Comment by non-Jedi** [comment]
> Interestingly, the LSP spec includes semantic tokens since 3.16. I wonder if that's supported in practice for Julia with Emacs, @gdkrmr and @non-Jedi, it would be great if you could share your thoughts about this. If we could make that work, it would cover pretty much everything for us.

LanguageServer.jl doesn't support the semantictokens set of capabilities at this time unfortunately. I'm also really not sure if the architecture of the LSP would give you sufficient responsiveness for indentation and syntax highlighting. Any time you press enter, emacs would have to make a round-trip with the language server (plaintext over pipes) before deciding on indentation level.

**Comment by tpapp** [comment]
I did not dig into the details, but I am still under the impression that tree-sitter would be the path of least resistance, because of:

1. existing Emacs integration,
2. existing Julia support, even if we need to fix things,
3. it is fast,
4. no need to spin up a Julia instance, so we could keep things reasonably simple in julia-emacs.

**Comment by non-Jedi** [comment]
I'm generally on board with integrating with tree-sitter. Even if support for julia syntax isn't perfect, it's probably better than what we have now especially wrt indentation.

We *would* need to update our lowest supported emacs version to 25.

**Comment by tpapp** [comment]
That's fine with me, Emacs 25 has been released almost 6 years ago.

**Comment by gdkrmr** [comment]
I don't really have any experience with syntax highlighting in Emacs, please correct me if I am wrong

- I don't think doing it through LanguageServer.jl/CSTParser.jl would be a good idea without a fallback. There is too much that can go wrong if you have to spin up Julia processes in the background. I also like the current modular approach: if for some reason `lsp-julia` fails, `julia-mode` is still very much usable (and users can switch to `eglot`). Including `CSTParser.jl` would probably mean that we have to put everything in a large monolith (why spin up two separate Julia processes if CSTParser and LanguageServer could run in a single process).
- tree-sitter sounds interesting and I don't mind increasing the minimum Emacs version to 25.
- tree-sitter uses dynamic modules which should be much more responsive than queries to an external Julia binary.

Questions:

- Would it be possible to wrap the Julia Language Server in a dynamic module?
- How does VSCode do syntax highlighting? Are they using CSTParser?

**Comment by ronisbr** [comment]
Is there anyway to enable tree-sitter in Julia mode? I installed, enabled, but it does not seem to have any effect.

**Comment by FelipeLema** [comment]
can you paste or point to the code you're dealing with?

**Comment by ronisbr** [comment]
Nothing in particular. I am just trying to see if we can have a better performance. For example, scrolling this file is somewhat slow for me:

https://github.com/ronisbr/PrettyTables.jl/blob/master/src/backends/text/print.jl

**Comment by gdkrmr** [comment]
emacs 29 will add native tree sitter support! Does anyone know how it works? Will there be an extra process or is it going to be a dynamic module? `tree-sitter-julia` also seems to be reasonably active, not sure if it is ready yet (https://github.com/tree-sitter/tree-sitter-julia).

**Comment by ronisbr** [comment]
To use tree-sitter, you need to rewrite the major mode. I am doing some experiment in a `julia-ts-mode` with very good outcome. There are rough edges, which I do not know yet if they are caused by emacs or julia grammar. I will publish this file tomorrow so people can test.

**Comment by ronisbr** [comment]
Here is the major mode:

https://github.com/ronisbr/julia-ts-mode

You need to add the file `julia-ts-mode.el` to you path and add `(require 'julia-ts-mode)`. Notice that you also need to install the Julia tree-sitter grammar.

I have to say that I am really amazed how easy it was to setup everything and the speed is definitely much faster than the current mode. Now, I need to work on navigation and imenu support.

**Comment by non-Jedi** [comment]
This is awesome. Thanks @ronisbr. I'll need to compile emacs 29 for myself to try this. Maybe you could instead define `julia-ts-mode` as a derived mode of `julia-mode` and only override `indent-line-function` and `font-lock-defaults` (that way we don't lose the pieces of julia-mode not related to indentation and font-locking)?

@tpapp would we be willing to make future releases of `julia-mode` only compatible with emacs 29+? This seems like a feature which would make doing so worthwhile.

**Comment by ronisbr** [comment]
Hi @non-Jedi !

> This is awesome. Thanks @ronisbr. I'll need to compile emacs 29 for myself to try this. Maybe you could instead define julia-ts-mode as a derived mode of julia-mode and only override indent-line-function and font-lock-defaults (that way we don't lose the pieces of julia-mode not related to indentation and font-locking)?

The idea is to test the tree-sitter integration and then commit to this repository. I will not register `julia-ts-mode`. Many major modes are defining something just as you said. Thus, the user can select if they want to old behavior or tree-sitter, if available. I think this is the best scenario.

> @tpapp would we be willing to make future releases of julia-mode only compatible with emacs 29+? This seems like a feature which would make doing so worthwhile.

Yes, probably we will need to require Emacs 29 to make this integration works.

**Comment by tpapp** [comment]
@non-Jedi: yes, definitely. @ronisbr: thanks for doing this. I believe that this is the best way to solve a long list of problems.

**Comment by ronisbr** [comment]
Perfect! I will ping this thread when I finish the initial version so that you can help me to integrate everything :)

**Comment by ronisbr** [comment]
Just an update:

I have been using Julia tree-sitter grammar for almost 3 weeks now. My doom configuration with this mode is here: https://github.com/ronisbr/doom.d/tree/emacs-29

Everything is working wonderfully! I found just two minor issues reported here:

https://github.com/tree-sitter/tree-sitter-julia/issues/88

https://github.com/tree-sitter/tree-sitter-julia/issues/73

The experience so far has been amazing.

**Comment by chriselrod** [comment]
eglot-jl currently [explicitly lists julia-mode](https://github.com/non-Jedi/eglot-jl/blob/07edf37eab5302c4ecca4edd1368855e9b5d0f73/eglot-jl.el#L10) as a dependency, but it seems like it'd be straightforward to give [julia-ts-mode](https://github.com/ronisbr/julia-ts-mode) a try. I'm not sure what I'd be looking for -- just see if I don't encounter problems, and maybe if it feels snappier?
I'm currently using c++-ts-mode, and maybe it adds more features, but I haven't tried them yet.
`mark-paragraph` still seems dumb / based on spacing rather than syntax.

```julia
function foo()

#mark begins
end
function bar()
# mark ends

end
```
Or maybe I need to look into documentation, and I'm supposed to replace functions like this with tree-sitter-powered versions.

**Comment by ronisbr** [comment]
> just see if I don't encounter problems, and maybe if it feels snappier?

Yes! 

> Or maybe I need to look into documentation, and I'm supposed to replace functions like mark-paragraph with tree-sitter-powered versions.

There is some support for navigation, but I did not change anything related to `mark-paragraph`.

**Comment by non-Jedi** [comment]
@chriselrod Looking at `eglot-jl` again, I don't think `julia-mode` is needed as a dependency. All that would be required to use it with `julia-ts-mode` would be to modify `eglot-server-programs` to instead include `'(julia-ts-mode . eglot-jl--ls-invocation)`.

@ronisbr I haven't had a chance to build emacs 29 and test this, but would you mind skimming through the open issues when you get a chance and seeing which ones would be solved by your `julia-ts-mode` implementation? I would like to test for at least #118, #111, #56, #12, #11 (especially #11!!) #3, #2. If you don't have time, that's understandable, and I'll take a look when I get a chance.

**Comment by chriselrod** [comment]
With this, eglot-jl should be compatible with both julia-mode and julia-ts-mode: https://github.com/non-Jedi/eglot-jl/pull/36
(I just swapped `julia-mode` for `(julia-mode julia-ts-mode)`).

Building emacs is fairly straightfoward.
On fedora, you can install all the dependencies via:
```sh
sudo dnf install -y xset xsetroot slock dnf-utils libgcc-jit-devel libtree-sitter-devel stow
sudo yum-builddeps emacs
```
Then to build
```sh
# cd somewhere/so/you/do/not/clutter
git clone git://git.savannah.gnu.org/emacs.git
cd emacs
./autogen.sh
mkdir build
cd build
CFLAGS="-O3 -march=native -fno-semantic-interposition" CXXFLAGS="-O3 -march=native -fno-semantic-interposition" ../configure --with-native-compilation --with-wide-int --with-json --with-tree-sitter
time make NATIVE_FULL_AOT=1 -j(nproc)
sudo make install prefix=/usr/local/stow/emacs
cd /usr/local/stow
sudo stow emacs
```
`stow` is nice so you can easily clear out something you've installed (`sudo stow -D emacs` will remove all the symlinks it creates).

You may want to change flags/configuration options/etc.
I obviously added native compilation and treesitter above.
You may want pure-GTK if you're using Wayland, but I need to build with X because I'm using EXWM.

**Comment by chriselrod** [comment]
https://github.com/JuliaEditorSupport/julia-emacs/issues/118
julia-ts-mode only highlights `quote` and `end`.

https://github.com/JuliaEditorSupport/julia-emacs/issues/111
New line after `quote` does not indent; typing without tab:
```julia
map(1:3) do x
x
end
f(map(1:3) do x
x
end)
```
mark and tab:
```julia
map(1:3) do x
  x
end
f(map(1:3) do x
  x
end)
```

https://github.com/JuliaEditorSupport/julia-emacs/issues/56
```julia
x .|>
  f
```
It initially didn't have the indent, but as soon as I made another line, it auto-indented.

https://github.com/JuliaEditorSupport/julia-emacs/issues/12
```julia
module A
import Base: *
a = 1
b = *
c = 2
end
```
is what I get typing it out (no spurious indent).
Mark and tab preserves the correct relative indent, except it indents everything inside the module.
The `module` keyword is not itself highlighted.
```julia
module A
  import Base: *
  a = 1
  b = *
  c = 2
end
```
https://github.com/JuliaEditorSupport/julia-emacs/issues/11
This behavior is customizable https://github.com/ronisbr/julia-ts-mode/blob/197e6e81a8d3d519df81fd21931a1e156ec1fc10/julia-ts-mode.el#L44-L84

Typing it out:
```julia
function1(a, b, c
  d, e, f)
function2(
a, b, c
  d, e, f)
for i in Float64[1, 2, 3, 4
  5, 6, 7, 8]
end
for i in Float64[
1, 2, 3, 4
  5, 6, 7, 8]
end
a = function3(function()
return 1
end)
a = function4(
function ()
return 1
end)
```
Neither leading `5` is highlighted, but all the other numbers are.
Mark and tab:
```julia
function1(a, b, c
  d, e, f)
function2(
  a, b, c
  d, e, f)
for i in Float64[1, 2, 3, 4
  5, 6, 7, 8]
end
for i in Float64[
  1, 2, 3, 4
  5, 6, 7, 8]
end
a = function3(function()
  return 1
end)
a = function4(
  function ()
    return 1
  end)
```

So, still some problems.
Seems like it is missing a lot of indents as you type (not the biggest deal -- I auto format on save), the two most visually annoying things to me are that the `5`s have the default face, unlike the rest of the numbers.
`in` and function calls don't have any highlighting either, I'll have to look if that's also customizable.

**Comment by chriselrod** [comment]
https://github.com/JuliaEditorSupport/julia-emacs/issues/3
No special highlighting for `user`
```julia
println("hello $user")
```

https://github.com/JuliaEditorSupport/julia-emacs/issues/2
No highlighting for any of these variables.

**Comment by chriselrod** [comment]
Just to confirm, with this file:
https://gist.github.com/chriselrod/5d09f5156ee49f1d2822df1638093b76#file-highssimplex-jl
`julia-mode` seems quite fast, while `julia-ts-mode` lags.

**Comment by ronisbr** [comment]
Thanks for the amazing investigation @chriselrod !

> https://github.com/JuliaEditorSupport/julia-emacs/issues/118
> julia-ts-mode only highlights quote and end.

Ooops :D I forgot to add anything related with interpolation expressions. Now we have:

![Captura de Tela 2023-01-10 às 10 11 41](https://user-images.githubusercontent.com/1068295/211560550-14781016-7924-4f93-acad-c60d26dbba79.png)

> -[ ] Not very eager about indenting new lines, but seems pretty good in terms of consistent indentation levels when you ask for them.

I saw problems like this also in other languages. It seems some limitation on either the Emacs implementation or tree sitter itself.
For example, when you type:

```
if a == 2
end
```

And press enter after `2`, the line is not indented. The reason is that there is no node inside the if statement. Hence, Emacs just does not know that we want to shift the indentation. I have no idea how to fix it. Perhaps https://github.com/tree-sitter/tree-sitter-julia/issues/73 will improve it.

> Neither leading 5 is highlighted, but all the other numbers are.

This happens because it is a syntax error. We can highlight errors. However, the grammar is not 100% and some errors are false positives.

> https://github.com/JuliaEditorSupport/julia-emacs/issues/3
> No special highlighting for user
> 
> `println("hello $user")`

Done! I added the support for string interpolations. The font face is the same as in the constant, but bold.

<img width="202" alt="Captura de Tela 2023-01-10 às 10 36 49" src="https://user-images.githubusercontent.com/1068295/211565757-32f256f3-97cd-42ac-aff5-ddda8b8e07ed.png">

> https://github.com/JuliaEditorSupport/julia-emacs/issues/2
> No highlighting for any of these variables.

I think I did not fully understand what is the desired behavior. Can you please explain to me?

> julia-ts-mode highlights the giant matrix, but julia-mode does not.
> Perhaps that is the cause of the performance difference?
> I'll have to try more normal files.

Yes! If you set `trees it-font-lock-level` to 2, where the literals are not highlighted, it is much faster.

**Comment by ronisbr** [comment]
> I'd like in in the loops to be highlighted, as well as of course the 5s.

The `in` highlighting is done! The `5` in your example is impossible because tree-sitter will mention that there is an error.

<img width="90" alt="Captura de Tela 2023-01-10 às 10 54 03" src="https://user-images.githubusercontent.com/1068295/211569499-d545eec2-1e54-42f1-8d71-a1ada3df8746.png">

**Comment by non-Jedi** [comment]
> > #2
> > No highlighting for any of these variables.
> 
> I think I did not fully understand what is the desired behavior. Can you please explain to me?

Any time a variable is assigned to, the variable name should be highlighted with `font-lock-variable-name-face`, e.g.:

- the `x` in `x = 5`
- `a` and `b` in `let a = 1, b = 2`
- `x` and `y` in `x, y = 4, 5`
- `a` in `global a`
- `b` in `local b`
- `i` in `for i=1:10` (this one is arguable)
- `x` and `y` and the `T` in the `where` clause in `function f(x::T, y) where T` (this one is arguable)

But there are similar forms which should not be highlighted which makes this difficult to do without the full parser we get with tree-sitter (e.g. named tuples and `setindex!` sugar).

**Comment by ronisbr** [comment]
Hi @non-Jedi ,

Thanks!

Everything that is not arguable was implemented. However, we might have corner cases.

I add the variable highlighting to the level 3 (the default). This is the specification for each level:

> Level 1 usually contains only comments and definitions.
> Level 2 usually adds keywords, strings, constants, types, etc.
> Level 3 usually represents a full-blown fontification, including
> assignment, constants, numbers, properties, etc.
> Level 4 adds everything else that can be fontified: delimiters,
> operators, brackets, all functions and variables, etc.

Now we have:

<img width="195" alt="Captura de Tela 2023-01-10 às 16 12 33" src="https://user-images.githubusercontent.com/1068295/211640691-8a3e9108-c83a-468c-9364-3ea0fc499225.png">

**Comment by chriselrod** [comment]
Thanks for all the great work!
(Of course, thanks to everyone developing the packages I use.)

> Yes! If you set `treesit-font-lock-level` to 2, where the literals are not highlighted, it is much faster.

For now I went in the other direction, and am trying `4`.
```elisp
(defun treesit-font-lock-level-4 ()
  (setq-local treesit-font-lock-level 4)
  (treesit-font-lock-recompute-features))
(add-hook 'julia-ts-mode-hook #'treesit-font-lock-level-4)
```

Something else I noticed: macros aren't highlighted.

**Comment by ronisbr** [comment]
> Thanks for all the great work!

You're welcome! I also want to point out the AMAZING work of Julia tree-sitter grammar developers (@maxbrunsfeld, @savq, and others). In all this time, I only found very minor issues! It is amazing!

> For now I went in the other direction, and am trying 4.

Me too! However, it can slow down. I noticed that the problem is when at the screen there is a lot of highlighting. It seems that it can handle big files pretty well (did not test deeply).

> Something else I noticed: macros aren't highlighted.

I did not understand, it seems to be working here:

<img width="143" alt="Captura de Tela 2023-01-11 às 09 27 04" src="https://user-images.githubusercontent.com/1068295/211806156-364b4858-8b2c-4732-b50c-c14aab36f4bf.png">

By the way, until `julia-ts-mode` is merged here, I needed to replicate some functionality. Hence, I copied all the code related with LaTeX symbol substitution. Now, I think it is working to the point I can start using it daily.

**Comment by chriselrod** [comment]
> I did not understand, it seems to be working here:

Hmm -- I'm on a different computer with the same config as before, and I now see the same thing you showed.
Perhaps I was on outdated versions. I'll let you know if I see anything different.

**Comment by ronisbr** [comment]
> By the way, until julia-ts-mode is merged here, I needed to replicate some functionality. Hence, I copied all the code related with LaTeX symbol substitution. Now, I think it is working to the point I can start using it daily.

I undid this given the amazing advice of @non-Jedi to make `Julia-ts-mode` a derived mode of `Julia-mode`.

**Comment by ronisbr** [comment]
Update!

After a lot of problems, I managed to make an option to select which kind of indentation after assignment the user wants. Hence, we can now select:

```julia
    var = a + b + c +
          d + e +
          f
```

or

```julia
    var = a + b + c +
        d + e +
        f
```

**Comment by tpapp** [comment]
I would like to revisit this issue.

@dhanak, @non-Jedi: as recent contributors to the julia-ts-mode, your opinion would be especially valuable.

Specifically, if tree-sitter works, or works better than current regexp-based highlighting, we could switch to that. FWIW, I am fine with requiring Emacs 29 if needed, as it is now almost a years old. (I don't know how the third-party package works).

**Comment by dhanak** [comment]
Hi, @tpapp, I've been using julia-ts-mode for about a year now, and I've spent quite some time tuning and updating its syntax highlighting rules. The version I'm using currently is available at https://github.com/dhanak/julia-ts-mode/, which is a few modifications ahead of the official version available from MELPA.

If you'd like to merge it with julia-mode, I'd be happy to assist you.

**Comment by tpapp** [comment]
@dhanak, thanks for the feedback. I have been using the official version and ran into an issue (https://github.com/JuliaEditorSupport/julia-ts-mode/issues/30) already, so I am waiting to see how quick these are fixed or even responded to.

I am curious why you are using a branch instead of the official version, is it because fixed are merged slowly?

Also, the way I imagined "integrating" tree-sitter is just loading and enabling it, and at the same time disabling our own code which duplicates functionality. Is there more to it?

**Comment by non-Jedi** [comment]
@tpapp I'm assuming the reason @dhanak is using his fork is that there's nobody with commit permissions really maintaining `julia-ts-mode` currently. I've helped him get some fixes merged in the past, but I haven't chosen to devote much time to that kind of thing lately, and I haven't been using `julia-ts-mode` myself. If @dhanak is willing to help maintain `julia-ts-mode`, we should ask to get him added to the JuliaEditorSupport org.

For the "how" of integrating: the approach for other major modes seems to be to have both the ts mode and regex-based modes available. Installing the treesitter grammar is not easily automated on Windows last I heard, so this seems like the approach we should take as well. There's some emacs machinery for switching between the two, as I recall, which lets the name `julia-mode` name point to either the current julia-mode or `julia-ts-mode` based on customization.

**Comment by dhanak** [comment]
I forked it because I made several improvements to the font locking related code. The first half of this got merged (thanks to @non-Jedi), but a second batch of further improvements (Julia 1.12 updates and GitLab CI integration) is still waiting in a [PR](https://github.com/JuliaEditorSupport/julia-ts-mode/pull/28).

Note that I wouldn't mind getting added to the JuliaEditorSupport org, albeit I have no significant experience with Emacs Lisp package maintenance.

Enabling tree sitter is only half of the story, the other half is adding treesit rules to connect the grammar with the font faces and indentation rules.

**Comment by tpapp** [comment]
I was thinking of addressing various parser/highlighter bugs, so I am reviving this topic to see if there have been new developments and think of a general strategy that would make sense in the long run. Please chime in everyone, even if you have a wild suggestion.

The "official" grammar repo for tree-sitter/julia seems to be quite stagnant, PRs not looked at for months, etc. This is in line with what @FelipeLema wrote above at the very beginning of this discussion.

On the Julia side, with app support, precompilation, using a Julia-based solution could be a possibility, if not now, then in the near future. But I am wondering eg how robust CSTParser.jl is when it comes to half-baked / incorrect code, which can happen during editing.

**Comment by FelipeLema** [comment]
> how robust CSTParser.jl is when it comes to half-baked / incorrect code, which can happen during editing

It's used in https://github.com/julia-vscode/LanguageServer.jl although I see that [incremental re-parsing is not implemented](https://github.com/julia-vscode/LanguageServer.jl/blob/a53652283fff7e54e5cf7821f7442af7f717137b/test/test_edit.jl#L46).

So I'm guessing "it's not fine-tuned for live editing, but It Mostly Works™".

I remember "being too slow" was never a problem with highlight when using LanguageServer.jl, although I wouldn't consider using LanguageServer.jl as a solution for highlighting due to the baggage of problems appearing in runtime.

NOTE: once you have tokens (type, begin, end... ) you can re-use code from [eglot's painting function for LSP's semantic tokens](https://github.com/joaotavora/eglot/blob/7084a59906bc7796cc369af4010500998140ab1c/eglot.el#L5098-L5121)

**Comment by tpapp** [comment]
After reading up about the options, I am inclined to try to get tree-sitter working as the first approach. Bozhidar Batsov [has a writeup](https://batsov.com/articles/2026/02/27/building-emacs-major-modes-with-treesitter-lessons-learned/) of his experience and the roadbumps with OCaml and Clojure, but they seem tolerable.

My proposed roadmap:

1. I approached @non-Jedi about adding @dhanak and I to the julia-ts-mode repo as maintainers, so we could merge commits. I consider [rust-ts-mode](https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/rust-ts-mode.el) a good example of best practices so I would follow what they do unless there is a reason not to.

2. I would propose forking the latest ts grammar, and working on that.
